### PR TITLE
Intelligent scissors with OpenCV javascript

### LINKED ISSUE #336: Intelligent scissors
Hey,

I am trying to extend the polygon and polyline tool to snap to a contour. This would make the annotation process for objects with a strong gradient more efficient. I am planning to implement a variation of "[Intelligent Scissors for Image Composition](https://www.cs.drexel.edu/~david/Classes/Papers/p191-mortensen.pdf)" by Eric N. Mortensen & William A. Barrett.

For this purpose I'd need access to the click event when adding/editing a point while in creation mode for polygons/polylines, as well as the gradient information of the image.
The later one I could pre-compute or calculate it once the "magnetic mode" is activated.

Does anyone have a hint for me which parts of the engine are crucial to achieve this?

Cheers,

Matt

### PR DESCRIPTION
**Pull request by bsekachev**
### Motivation and context
Integration of OpenCV.js library into CVAT. Now Intelligent Scissors have been added (resolved #336), but we expect more functions in the future.

Threshold maintenance using ```ctrl + mousewheel```
Screenshot from 2021-01-20 13-55-29
Screenshot from 2021-01-20 13-55-52

### How has this been tested?
Manual testing

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

### CODE REVIEW

**Comment by nmanovic** [comment]
@ActiveChooN , could you please review?

**Comment by bsekachev** [comment]
OpenCV PR: https://github.com/opencv/opencv/pull/19194
Demo: https://pullrequest.opencv.org/buildbot/export/pr/19194/docs/d9/df5/tutorial_js_intelligent_scissors.html

File: cvat-ui/src/utils/opencv-wrapper/intelligent-scissors.ts [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+            tool: new cv.segmentation_IntelligentScissorsMB(),
+            state: {
+                path: [],
+                anchors: {},
+                image: null,
+            },
+        };
+
+        this.scissors.tool.setEdgeFeatureCannyParameters(32, 100);
+        this.scissors.tool.setGradientMagnitudeMaxLimit(200);
```
Comment (alalek) [inline cvat-ui/src/utils/opencv-wrapper/intelligent-scissors.ts:65]: Looks like `reset()` below doesn't reapply these settings

**Review by alalek — COMMENTED** [review]
Great!
/cc @dkurt @alvoron @SimpleVlad

---

> cvat/apps/opencv/static/opencv/js/opencv.js

It would be nice to store somewhere some instructions how to update this file in the future

File: cvat-ui/src/utils/opencv-wrapper/intelligent-scissors.ts (hunk 1 above)
Comment (bsekachev) [reply cvat-ui/src/utils/opencv-wrapper/intelligent-scissors.ts]: Thx, fixed. Now it works better

File: cvat/apps/opencv/views.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import os
+from django.conf import settings
+from sendfile import sendfile
+
+def OpenCVLibrary(request):
+    path = os.path.realpath(os.path.join(settings.STATIC_ROOT, 'opencv', 'js', 'opencv.js'))
+    return sendfile(request, path)
```
Comment (nmanovic) [inline cvat/apps/opencv/views.py:7]: I like the solution with a separate app for OpenCV.js. Great job!

Comment (nmanovic) [reply cvat/apps/opencv/views.py:7]: Probably you want to include version information into the filename. For example, opencv-X.Y.Z.js. The file will be accessed using opencv.js but in repo we will see which version we have.

**Comment by nmanovic** [comment]
> Great!
> /cc @dkurt @alvoron @SimpleVlad
> 
> > cvat/apps/opencv/static/opencv/js/opencv.js
> 
> It would be nice to store somewhere some instructions how to update this file in the future

@alalek , do you think we can place information about the integration somewhere on OpenCV site? For example, tutorial or demo page? Let's say a statement like "the feature is integrated into [CVAT](https://github.com/openvinotoolkit/cvat) annotation tool and you can try it online on cvat.org".

File: cvat-ui/src/utils/opencv-wrapper/opencv-wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    private initialized: boolean;
+    private cv: any;
+
+    public constructor() {
+        this.initialized = false;
+        this.cv = null;
+    }
+
+    public async initialize(onProgress: (percent: number) => void): Promise<void> {
+        const response = await fetch(`${baseURL}/opencv/opencv.js`);
```
Comment (nmanovic) [inline cvat-ui/src/utils/opencv-wrapper/opencv-wrapper.ts:27]: Should we consider it as server API and add into server-proxy?

File: cvat/apps/opencv/views.py (hunk 2 above)
Comment (alalek) [reply cvat/apps/opencv/views.py:7]: Please note that related OpenCV code is not released yet (it is missing in 4.5.1, 4.5.2 is not released yet)

Comment (nmanovic) [reply cvat/apps/opencv/views.py:7]: Better to have something in any case. Probably 4.5.1 is better than nothing.

**Comment by nmanovic** [comment]
@bsekachev , build bs/opencv and got the following error:

image

**Comment by nmanovic** [comment]
```
commit b20f5028dfa74b8e63bc7a4180aa064e5ed4dda7 (HEAD -> bs/opencv, origin/bs/opencv)
Author: Boris Sekachev <boris.sekachev@intel.com>
Date:   Thu Jan 28 12:22:59 2021 +0300

    Fixed corner case
```

**Comment by ActiveChooN** [comment]
Most works really fine, but have some thoughts.

1. If we have no script on related path, click on load button will be completed with success icon.
2. Threshold can be configurated with consts or state for larger working area for example.
3.  Sometimes i've got an error message when move pointer fast enough. It quite rare and i connot record this case.
4. Guess we can improve interactivity by starting interactor if only previous interaction has finished. Without it you can only achieve interaction result when mouse pointer stops moving.

File: cvat/apps/opencv/views.py (hunk 2 above)
Comment (alalek) [reply cvat/apps/opencv/views.py:7]: Alternative is using package/PR creation date (20210120 or without day 202101):
`opencv-cvat-<YYYYmmdd>.js`

Comment (bsekachev) [reply cvat/apps/opencv/views.py:10]: Done

**Comment by bsekachev** [comment]
> @bsekachev , build bs/opencv and got the following error:

@nmanovic  Forgot to setup server & proxy, fixed.

**Comment by bsekachev** [comment]
@ActiveChooN 

> If we have no script on related path, click on load button will be completed with success icon.

Fixed. 

> Threshold can be configurated with consts or state for larger working area for example.

Ideally it should be configurated in UI, but it would be difficult to unify UI for all possible thresholds with different tools from OpenCV. Now we also do not have thresholds for automatic annotation, maybe it will be done in the future. 

> Guess we can improve interactivity by starting interactor if only previous interaction has finished. Without it you can only achieve interaction result when mouse pointer stops moving.

I am not sure I understand you correct

**Comment by nmanovic** [comment]
@bsekachev , an alternative way to implement the feature https://www.photoshopessentials.com/basics/selections/magnetic-lasso-tool/. What do you think?

Looks like it is a polygon feature with "shift" pressed but you don't need move the mouse along the boundary precisely.

File: cvat-ui/src/utils/opencv-wrapper/opencv-wrapper.ts (hunk 3 above)
Comment (nmanovic) [reply cvat-ui/src/utils/opencv-wrapper/opencv-wrapper.ts:27]: ?

**Review by nmanovic — APPROVED** [review]
Server part looks good for me. I think we can merge it as is for now.

**Comment by bsekachev** [comment]
Okay, let's merge it as an initial version and look how we can improve it in the future patches.

**Comment by jahaniam** [comment]
@bsekachev @nmanovic  Hi guys,
I've tested this feature. It looks great although it mostly works well when there is a clear boundary between background and foreground.

I have these two suggestions:
1. removing points is done using ctrl + clicking. It is not consistent with normal polygon drawing which is right-clicking. It would have been nicer to be consistent.
2. It would have been great to automatically draw a point when it is at the end of the rectangle (like how magnetic lasso tool does in photoshop. like how it is done in current cvat when holding shift. 

Thanks again for your work.

**Comment by bsekachev** [comment]
@jahaniam 

Hi, thanks for your feedback. Please, address your comment to dedicated issue #2800

**Pull request merged**